File: version.old/pp.first/java/try1/machine/Tape.java
package machine;

import java.net.*;
import java.io.*;
import java.util.*;

/**
 *  
 *  This class represents an infinite array of text strings, where each
 *  array cell can, in theory hold an infinite amount of text. The tape
 *  has a 'pointer' to keep track of the current cell in the array and 
 *  to allow the tape to be accessed without any parameters. For example
 *   get; will obtain the tape cell which is currently pointed to.
 *
 *  @author http://bumble.sf.net
 */
 
  
public class Tape extends Object
{
  //--------------------------------------------
  private static String NEWLINE = System.getProperty("line.separator");
  //--------------------------------------------
  private Vector array;
  //--------------------------------------------
  int pointer;



  //--------------------------------------------
  public Tape()
  {
    this.array = new Vector();
    this.pointer = 0;
  }


  //--------------------------------------------
  /** 
   *   */
  public String toString()
  {
    return this.array.toString();
    /*
    StringBuffer sbReturn = new StringBuffer();
    String currentItem = new String();
    Enumeration ii = this.array.elements();
    while (ii.hasMoreElements())
    {
      currentItem = (String)ii.nextElement();
      sbReturn.append("|" + currentItem);
    }
    return sbReturn.toString();
    */
  } //-- method:

  //--------------------------------------------
  /** 
   *   */
  public String print()
  {
    StringBuffer sbReturn = new StringBuffer();
    String currentItem = new String();
    
    sbReturn.append("Pointer==" + this.pointer + NEWLINE);
    for (int ii = 0; ii < array.size(); ii++)
    {
      currentItem = (String)array.elementAt(ii);
      sbReturn.append(ii + ":" + currentItem + NEWLINE);
    }
    return sbReturn.toString();
    
  } //-- method:

  //--------------------------------------------
  /** 
   *   */
  public void put(String sNewElement)
  {
    if (this.pointer == this.array.size())
    {
      this.array.addElement(sNewElement);
      return;
    }

    this.array.setElementAt(sNewElement, this.pointer);
  } //-- method:

  
  //--------------------------------------------
  /** 
   *   */
  public String get()
  {
    if (this.pointer == this.array.size())
     { return ""; }
    return (String)this.array.elementAt(this.pointer);
  } //-- method:

  //--------------------------------------------
  /** 
   *   */
  public void incrementPointer()
  {
    if (this.pointer == this.array.size())
    {
      this.array.addElement("");
    }

    this.pointer++;
  } //-- method:

  //--------------------------------------------
  /** 
   *   */
  public void decrementPointer()
  {
    if (this.pointer == 0)
     { return; }

    this.pointer--;
  } //-- method:


  //--------------------------------------------
  /** a main method for testing */
  public static void main(String[] args) throws Exception
  {
    
    StringBuffer sbUsageMessage = new StringBuffer("");
    sbUsageMessage.append("test usage: java Tape ");
    sbUsageMessage.append(NEWLINE);

    StringBuffer sbMessage = new StringBuffer("");


    if (args.length > 2)
    {	    
      System.out.println(sbUsageMessage);
      System.exit(-1);
    }

    //String sText = args[0];
    //char cChar = args[1].charAt(0);


    Tape testTape = new Tape();

    System.out.println("put 'diffle'               ");
    testTape.put("diffle");
    System.out.println(testTape);

    System.out.println("put 'buff'               ");
    testTape.put("buff");
    System.out.println(testTape);

    System.out.println(".incrementPointer()       ");
    testTape.incrementPointer();
    System.out.println(testTape);

    System.out.println("put 'diffle'               ");
    testTape.put("diffle");
    System.out.println(testTape);



  } //-- main()
  
} //-- Tape class

File: version.old/pp.first/cpp/Tape.cpp
//string example
//http://www.yolinux.com/TUTORIALS/LinuxTutorialC++StringClass.html
/*
 *  
 *  Is a virtual tape 
 *
 *  @author http://bumble.sf.net
 */

// good examples
// http://www.josuttis.com/libbook/i18n/loc1.cpp.html

#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include "Tape.h"

using namespace std;
  

  //--------------------------------------------
  Tape::Tape() 
  {
    vector<string> array;
    pointer = 0;
  }


  //--------------------------------------------
  string Tape::toString()
  {
    string sReturn("");
    stringstream ssReturn;
    string currentItem("");
    
    ssReturn << "Pointer==" << this->pointer << endl;
    for (int ii = 0; ii < array.size(); ii++)
    {
      currentItem = this->array[ii];
      ssReturn << ii << ":" << currentItem << endl;
    }
    
    sReturn = ssReturn.str();
    return sReturn;
  } //-- method:

  //--------------------------------------------
  string Tape::print()
  {
    string sReturn("");
    stringstream ssReturn;
    string currentItem("");
    
    ssReturn << "Pointer==" << this->pointer << endl;

    for (int ii = 0; ii < array.size(); ii++)
    {
      currentItem = this->array[ii];
      ssReturn << ii << ":" << currentItem << endl;
      //cout << ii << ":" << currentItem << endl;
    }
    
    return ssReturn.str();
    //cout << ssReturn.str();
  } //-- method:



  //--------------------------------------------
  void Tape::put(string sNewElement)
  {
    if (this->pointer == this->array.size())
    {
      this->array.push_back(sNewElement);
      return;
    }

    this->array[this->pointer] = sNewElement;
  } //-- method:

  
  //--------------------------------------------
  string Tape::get()
  {
    if (this->pointer == this->array.size())
     { return ""; }
    return this->array[this->pointer];
  } //-- method:

  //--------------------------------------------
  void Tape::incrementPointer()
  {
    if (this->pointer == this->array.size())
    {
      this->array.push_back("");
    }

    this->pointer++;
  } //-- method:

  //--------------------------------------------
  /*  */
  void Tape::decrementPointer()
  {
    if (this->pointer == 0)
     { return; }

    this->pointer--;
  } //-- method:


  //--------------------------------------------
  /* a main method for testing */
/*
  int main()
  {
    
    stringstream ssUsageMessage;
    ssUsageMessage << "test usage: java Tape ";

    string sMessage("");


    //String sText = args[0];
    //char cChar = args[1].charAt(0);


    Tape testTape;

    cout << "put 'diffle'               " << endl;
    testTape.put("diffle");
    cout << testTape.toString() << endl;

    cout << "put 'buff'               " << endl;
    testTape.put("buff");
    cout << testTape.toString() << endl;

    cout << ".incrementPointer()       " << endl;
    testTape.incrementPointer();
    cout << testTape.toString() << endl;

    cout << "put 'diffle'               " << endl;
    testTape.put("diffle");
    testTape.print();
    cout << testTape.toString() << endl;



  } //-- main()

*/
